src/freenas/usr/local/sbin/disk_resize

```diff
@@ -47,151 +47,105 @@ fi
+	block=`echo "${ident}" | grep -c "\*	BLOCK_ERASE_EXT command"`
+	crypto=`echo "${ident}" | grep -c "\*	CRYPTO_SCRAMBLE_EXT command"`
+	security=`echo "${ident}" | grep -c " for SECURITY ERASE UNIT"`
 	if [ ${security} -ne 0 ]; then
-		sfrozen=`camcontrol security ${dev} | grep -c "security config frozen *yes"`
-		if [  ${sfrozen} -ne 0 ]; then
+		snotfrozen=`echo "${ident}" | grep -c "not	frozen"`
+		if [ ${snotfrozen} -eq 0 ]; then
 			security="0"
 		fi
 	fi
-	trim=`diskinfo -v /dev/${dev} | grep -c "Yes.*# TRIM/UNMAP support"`
+	trim=`echo "${ident}" | grep -c "\*	Data Set Management TRIM supported"`
+	maxsect=`hdparm -N /dev/${dev} | awk -F '[/,]' '/max sectors/ { print $2 }'`
 
 	# When resizing SSD to specified size (down?) deallocate the flash.
 	if [ ${ssd} -ne 0 -a -n "${size}" ]; then
 		err=1
 		if [ ${block} -ne 0 -a ${err} -ne 0 ]; then
 			echo "Doing block erase sanitize."
-			camcontrol sanitize ${dev} -a block -y
+			hdparm --yes-i-know-what-i-am-doing --sanitize-block-erase /dev/${dev}
 			err=$?
 		fi
 		if [ ${crypto} -ne 0 -a ${err} -ne 0 ]; then
 			echo "Doing cryptograhic erase sanitize."
-			camcontrol sanitize ${dev} -a crypto -y
+			hdparm --yes-i-know-what-i-am-doing --sanitize-crypto-scramble /dev/${dev}
 			err=$?
 		fi
+		if [ ${err} -eq 0 ]; then
+			echo "Waiting for sanitize to complete."
+			until hdparm --sanitize-status /dev/${dev} | grep -Fq "SD0 Sanitize Idle"
+			do
+				sleep 10
+			done
+		fi
 		if [ ${security} -ne 0 -a ${err} -ne 0 ]; then
 			echo "Doing security erase."
-			camcontrol security ${dev} -q -s MyPass -e MyPass -y
+			hdparm --yes-i-know-what-i-am-doing --security-erase MyPass /dev/${dev}
 			err=$?
 		fi
 		if [ ${trim} -ne 0 -a ${err} -ne 0 ]; then
 			echo "Doing TRIM."
-			trim -fq /dev/${dev}
+			blkdiscard /dev/${dev}
 			err=$?
 		fi
 		if [ ${err} -ne 0 ]; then
 			echo "No method found to deallocate the flash."
 		fi
 	fi
 
-	if [ ${ama} -ne 0 ]; then
-		if [ -z "${size}" ]; then
-			size=`echo "${ident}" | awk -F "[ \t/]+" '/Accessible Max Address Config/ { print $8 }'`
-		fi
-		echo "Setting Accessible Max Address to ${size}"
-		camcontrol ama ${dev} -qs ${size}
-		err=$?
-	else
-		if [ -z "${size}" ]; then
-			size=`echo "${ident}" | awk -F "[ \t/]+" '/Host Protected Area/ { print $8 }'`
-		fi
-		echo "Setting Host Protected Area to ${size}"
-		camcontrol hpa ${dev} -qPy -s ${size}
-		err=$?
+	if [ -z "${size}" ]; then
+		size=${maxsect}
 	fi
-	if [ ${err} -eq 0 ]; then
-		if [ ${devtype} = "satl" ]; then
-			echo "Resetting device"
-			camcontrol reset ${dev}
-			echo "Resize completed successfully.  Reboot may be needed."
-		else
-			echo "Resize completed successfully."
-		fi
+	echo "Setting Max Sectors to ${size}"
+	hdparm --yes-i-know-what-i-am-doing -N p${size} /dev/${dev}	# NB: the p is for persistence
+	if [ $? -eq 0 ]; then
+		echo "Resize completed successfully.  Reboot may be needed."
```

Comment:
I tried hdparm -w for device reset but it didn't work on anything I tested it on and it's marked as (DANGEROUS) so I figured better safe than sorry and took it out.

Reply:
It could be investigated deeper later.  It would be good if most typical cases could avoid reboot. Otherwise resize support in middleware is not very useful.